sklearn/metrics/pairwise.py

```diff
@@ -671,6 +686,260 @@ def cosine_distances(X, Y=None):
+        raise TypeError("Gower distance does not support sparse matrices")
+
+    if not isinstance(scale, (bool, list, np.ndarray)):
+        raise TypeError("Parameter scale must be boolean, list, or ndarray")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    y_is_not_none = Y is not None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    # as np.object, otherwise numeric columns will be converted to string
+    # if there are other string columns.
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X, dtype=np.object)
+
+    if y_is_not_none and not isinstance(Y, np.ndarray):
+        Y = np.asarray(Y, dtype=np.object)
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=X.dtype,
+                                 force_all_finite=False)
+
+    n_rows, n_cols = X.shape
+
+    categorical_features_obj, categorical_features_num, numerical_features = \
+        _detect_categorical_features(X, categorical_features)
+
+    # Categorical columns
+    X_cat_obj = X[:, categorical_features_obj].astype(np.object)
+    X_cat_num = X[:, categorical_features_num].astype(np.float)
+
+    # Numerical columns
+    X_num = X[:, numerical_features].astype(np.float)
+
+    Y_cat_obj = None
+    Y_cat_num = None
+    Y_num = None
+    if y_is_not_none:
+        Y_cat_obj = Y[:, categorical_features_obj].astype(np.object)
+        Y_cat_num = Y[:, categorical_features_num].astype(np.float)
+        Y_num = Y[:, numerical_features].astype(np.float)
+    else:
+        Y_cat_obj = X_cat_obj
+        Y_cat_num = X_cat_num
+        Y_num = X_num
+
+    num_cols = X_num.shape[1]
+    n_col_num_present = num_cols > 0
+    n_col_cat_obj_present = X_cat_obj.shape[1] > 0
+    n_col_cat_num_present = X_cat_num.shape[1] > 0
+
+    # Calculates the min and max values, and if requested, scale the
+    # input values in order to obtain the distances between 0 and 1,
+    # as proposed by the Gower's paper.
+    ranges = None
+    if n_col_num_present:
+        process_scale = False
+        if isinstance(scale, bool):
+            process_scale = scale
+        else:
+            if (isinstance(scale, list) and len(scale) != X_num.shape[1]) or\
+               (isinstance(scale, np.ndarray) and
+                    len(scale.flat) != X_num.shape[1]):
+                raise ValueError("Length of scale parameter must be equal "
+                                 "to the number of numerical columns.")
+            process_scale = True
+
+        ranges, min, max = _precompute_gower_params(X_num, Y_num, scale)
+
+        if not process_scale and (np.min(min) < 0 or np.max(max) > 1):
+            raise ValueError("Input data is not scaled between 0 and 1.")
+
+    y_n_rows = Y.shape[0]
+    D = np.zeros((n_rows, y_n_rows), dtype=np.float)
+
+    for i in range(n_rows):
+        j_start = i
+
+        # for non square results
+        if n_rows != y_n_rows or y_is_not_none:
```

Comment:
After check pairwise arrays X will be assigned to Y if Y is initially None, Y is X won't work here and other points. 

This `if n_rows != y_n_rows or y_is_not_none` is to check if we need to adjust i for a non squared iteration. For this it must have different number of rows and Y must be passed as parameter since the beginning, that won't be the case if we test Y is not X.

We already discussed this before, I can retrieve the conversation in the loop if needed.

Reply:
Double checking this...

Reply:
The X is Y works here, but if nobody objects I prefer to use the y_is_not_none flag for performance sake, because this works inside a repetitive loop, and the flag is ~20% faster than to use X is Y. The code below proves this.

`import numpy as np
import time
X = np.array([1,2,3])
Y = np.array([3,4,5])
elapsed_sum_xisy = 0
elapsed_sum_flag = 0
for j in range(0, 10):
    sum_elapsed = 0
    for i in range(0, 1000000):
        start = time.time()
        if X is Y: pass
        sum_elapsed = sum_elapsed + (time.time() - start)

    elapsed_sum_xisy = elapsed_sum_xisy + sum_elapsed
    sum_elapsed =0
    flag = X is Y
    for i in range(0, 1000000):
        start = time.time()
        if flag: pass;
        sum_elapsed = sum_elapsed + (time.time() - start)

    elapsed_sum_flag = elapsed_sum_flag + sum_elapsed

print('avg x is y elapsed:',elapsed_sum_xisy/10)
print('avg elapsed flag:',elapsed_sum_flag/10)
print('% gain of flag:', (1-elapsed_sum_flag/elapsed_sum_xisy) * 100)
`